File: solutions/cpp/Elias_coding.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <iostream>
#include <iterator>
#include <random>
#include <string>
#include <vector>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::random_device;
using std::string;
using std::uniform_int_distribution;
using std::vector;

string TransIntToBinary(int decimal);
int TransBinaryToInt(const string& binary);

// @include
string Encode(const vector<int>& A) {
  string ret;
  for (int a : A) {
    string binary = TransIntToBinary(a);
    binary.insert(0, binary.size() - 1, '0');  // prepend 0s.
    ret += binary;
  }
  return ret;
}

string TransIntToBinary(int decimal) {
  string ret;
  while (decimal) {
    ret.push_back('0' + (decimal % 2));
    decimal >>= 1;
  }
  reverse(ret.begin(), ret.end());
  return ret;
}

vector<int> Decode(const string& s) {
  vector<int> ret;
  int idx = 0;
  while (idx < s.size()) {
    // Count the number of consecutive 0s.
    int zero_idx = idx;
    while (zero_idx < s.size() && s[zero_idx] == '0') {
      ++zero_idx;
    }

    int len = zero_idx - idx + 1;
    ret.emplace_back(TransBinaryToInt(s.substr(zero_idx, len)));
    idx = zero_idx + len;
  }
  return ret;
}

int TransBinaryToInt(const string& binary) {
  int ret = 0;
  for (char c : binary) {
    ret = (ret * 2) + c - '0';
  }
  return ret;
}
// @exclude

int main(int argc, char* argv[]) {
  vector<int> A;
  default_random_engine gen((random_device())());
  if (argc == 1) {
    generate_n(back_inserter(A), uniform_int_distribution<int>(1, 10000)(gen),
               [&] { return uniform_int_distribution<int>()(gen); });
  } else {
    generate_n(back_inserter(A), atoi(argv[1]),
               [&] { return uniform_int_distribution<int>(4, 4)(gen); });
  }
  string ret = Encode(A);
  cout << ret << endl;

  vector<int> res = Decode(ret);
  assert(A.size() == res.size());
  for (int i = 0; i < A.size(); ++i) {
    assert(res[i] == A[i]);
  }
  return 0;
}

File: solutions/python/elias_coding.py
# Elias_coding.cpp 848813e190b1b85a8e75107fe8513c3be38ad1a9
import sys
import random


# @include
def encode(A):
    ret = []
    for a in A:
        binary = trans_int_to_binary(a)
        binary = '0' * (len(binary) - 1) + binary  # prepend 0s.
        ret.append(binary)
    return ''.join(ret)


def trans_int_to_binary(decimal):
    ret = []
    while decimal:
        ret.append(chr(ord('0') + (decimal & 1)))
        decimal >>= 1
    return ''.join(reversed(ret))


# @exclude


# Pythonic solution
def trans_int_to_binary_pythonic(decimal):
    return bin(decimal)[2:]


# @include


def decode(s):
    ret = []
    idx = 0
    while idx < len(s):
        # Count the number of consecutive 0s.
        zero_idx = idx
        while zero_idx < len(s) and s[zero_idx] == '0':
            zero_idx += 1

        length = zero_idx - idx + 1
        ret.append(trans_binary_to_int(s[zero_idx:zero_idx + length]))
        idx = zero_idx + length
    return ret


def trans_binary_to_int(binary):
    ret = 0
    for c in binary:
        ret = (ret * 2) + ord(c) - ord('0')
    return ret


# @exclude


# Pythonic solution
def trans_binary_to_int_pythonic(binary):
    return int(binary, 2)


def main():
    if len(sys.argv) == 1:
        A = [
            random.randint(1, (1 << random.randint(1, 31)) - 1)
            for i in range(random.randint(1, 10000))
        ]
    else:
        A = [
            random.randint(1, (1 << random.randint(1, 31)) - 1)
            for i in range(int(sys.argv[1]))
        ]

    ret = encode(A)
    print(ret)

    res = decode(ret)
    assert A == res


if __name__ == '__main__':
    main()
